runtime/stdlib/test.go

```diff
@@ -1367,6 +1428,280 @@ var equalMatcherFunction = interpreter.NewUnmeteredHostFunctionValue(
+		Parameters:           []sema.Parameter{},
+		ReturnTypeAnnotation: sema.NewTypeAnnotation(matcherType),
+	}
+}()
+
+var beEmptyMatcherFunction = interpreter.NewUnmeteredHostFunctionValue(
+	beEmptyMatcherFunctionType,
+	func(invocation interpreter.Invocation) interpreter.Value {
+		beEmptyTestFunc := interpreter.NewHostFunctionValue(
+			nil,
+			matcherTestFunctionType,
+			func(invocation interpreter.Invocation) interpreter.Value {
+				var isEmpty bool
+				switch value := invocation.Arguments[0].(type) {
+				case *interpreter.ArrayValue:
+					isEmpty = value.Count() == 0
+				case *interpreter.DictionaryValue:
+					isEmpty = value.Count() == 0
+				default:
+					panic(errors.NewDefaultUserError("expected Array or Dictionary argument"))
+				}
+
+				return interpreter.AsBoolValue(isEmpty)
+			},
+		)
+
+		return newMatcherWithGenericTestFunction(invocation, beEmptyTestFunc)
+	},
+)
+
+const haveElementCountMatcherFunctionName = "haveElementCount"
+
+const haveElementCountMatcherFunctionDocString = `
+Returns a matcher that succeeds if the tested value is an array or dictionary,
+and has the given number of elements.
+`
+
+var haveElementCountMatcherFunctionType = func() *sema.FunctionType {
+	return &sema.FunctionType{
+		IsConstructor:  false,
+		TypeParameters: []*sema.TypeParameter{},
+		Parameters: []sema.Parameter{
+			{
+				Label:      sema.ArgumentLabelNotRequired,
+				Identifier: "count",
+				TypeAnnotation: sema.NewTypeAnnotation(
+					sema.IntType,
+				),
+			},
+		},
+		ReturnTypeAnnotation: sema.NewTypeAnnotation(matcherType),
+	}
+}()
+
+var haveElementCountMatcherFunction = interpreter.NewUnmeteredHostFunctionValue(
+	haveElementCountMatcherFunctionType,
+	func(invocation interpreter.Invocation) interpreter.Value {
+		count, ok := invocation.Arguments[0].(interpreter.IntValue)
+		if !ok {
+			panic(errors.NewUnreachableError())
+		}
+
+		haveElementCountTestFunc := interpreter.NewHostFunctionValue(
+			nil,
+			matcherTestFunctionType,
+			func(invocation interpreter.Invocation) interpreter.Value {
+				var matchingCount bool
+				switch value := invocation.Arguments[0].(type) {
+				case *interpreter.ArrayValue:
+					matchingCount = value.Count() == count.ToInt(invocation.LocationRange)
+				case *interpreter.DictionaryValue:
+					matchingCount = value.Count() == count.ToInt(invocation.LocationRange)
+				default:
+					panic(errors.NewDefaultUserError("expected Array or Dictionary argument"))
+				}
+
+				return interpreter.AsBoolValue(matchingCount)
+			},
+		)
+
+		return newMatcherWithGenericTestFunction(invocation, haveElementCountTestFunc)
+	},
+)
+
+const containMatcherFunctionName = "contain"
+
+const containMatcherFunctionDocString = `
+Returns a matcher that succeeds if the tested value is an array that contains
+a value that is equal to the given value, or the tested value is a dictionary
+that contains an entry where the value is equal to the given value.
```

Comment:
For dictionaries, the method description says "..contains an entry where the value is equal to the given value", whereas the implementation checks for the existence of the key (i.e: `value.ContainsKey()` ). Maybe update the description to be consistent with the implementation.

```suggestion
that contains an entry where the key is equal to the given value.
```

Reply:
Good catch :100: . This got me thinking, which functionality would be more valuable to developers, checking for a `key` or `value`, or both? What is your opinion about it?

Reply:
Checking for the existence of keys is the most common, I guess. So the current implementation makes sense to me.

If someone wants to check for a value, they can get the values as an array using `value.values` and then pass it to the matcher.

Reply:
Perfect, thanks!